### ISSUE TITLE
#![allow(unknown_lints)] does not work if there is a unknow lint at the same level since nightly-2017-08-11-x86_64-unknown-linux-gnu

### ISSUE THREAD

**Issue by overvenus** [rust-lang/rust#43809]
The following snippet works fine before `nightly-2017-08-11-x86_64-unknown-linux-gnu`

```rust
#![allow(unknown_lints)]
#![allow(clippy)]
```

but now it reports a warning: 

```
warning: unknown lint: `clippy`
 --> /home/poc/src/../a/b.rs:2:10
  |
2 | #![allow(clippy)]
  |          ^^^^^^
  |
  = note: #[warn(unknown_lints)] on by default
``` 

Steps to reproduce: 

 1. clone https://github.com/overvenus/poc.git
 2. rustup override set nightly-2017-08-11
 3. cargo test

Is this an expected behavior, or a bug? Thank you!

**Comment by Mark-Simulacrum** [comment]
cc @alexcrichton -- probably related to the new lint implementation

**Comment by alexcrichton** [comment]
Yes this was caused by https://github.com/rust-lang/rust/pull/43522. The bug is on [this line of code](https://github.com/alexcrichton/rust/blob/0374e6aab7ef60b523872556ae4aca33c59fbfc9/src/librustc/lint/levels.rs#L225-L228). The check for whether `allow_lints` is in scope is not checking the current set of attributes being linted.

For example this does not warn:

```rust
#![allow(unknown_lints)]

#[allow(clippy)]
fn main() {}
```

but this warns:

```rust
#![allow(unknown_lints)]
#![allow(clippy)]

fn main() {}
```

(as you've noticed)

The fix is to just take the local variable `specs` into account which has the up-to-that-point set of lint attributes collected.

**Comment by alexcrichton** [comment]
I've opened a PR for this at https://github.com/rust-lang/rust/pull/43841
